File: frontend/services/productService.ts
import api from './api';
import { Product } from '@/types';

/**
 * Servicio de productos
 */
export const productService = {
  /**
   * Obtiene todos los productos disponibles (público)
   */
  async getAvailableProducts(): Promise<Product[]> {
    const response = await api.get<Product[]>('/products/public/all');
    return response.data;
  },

  /**
   * Obtiene un producto por ID
   */
  async getProductById(id: number): Promise<Product> {
    const response = await api.get<Product>(`/products/public/${id}`);
    return response.data;
  },

  /**
   * Obtiene todos los productos (incluye no disponibles) - ADMIN
   */
  async getAllProducts(): Promise<Product[]> {
    const response = await api.get<Product[]>('/products/admin/all');
    return response.data;
  },

  /**
   * Crea un nuevo producto - ADMIN
   */
  async createProduct(product: Omit<Product, 'id'>): Promise<Product> {
    const response = await api.post<Product>('/products/admin', product);
    return response.data;
  },

  /**
   * Actualiza un producto - ADMIN
   */
  async updateProduct(id: number, product: Partial<Product>): Promise<Product> {
    const response = await api.put<Product>(`/products/admin/${id}`, product);
    return response.data;
  },

  /**
   * Elimina un producto - ADMIN
   */
  async deleteProduct(id: number): Promise<void> {
    await api.delete(`/products/admin/${id}`);
  },

  /**
   * Cambia la disponibilidad de un producto - ADMIN
   */
  async toggleAvailability(id: number): Promise<Product> {
    const response = await api.patch<Product>(`/products/admin/${id}/toggle-availability`);
    return response.data;
  },

  /**
   * Busca productos por categoría
   */
  async getProductsByCategory(categoria: string): Promise<Product[]> {
    const response = await api.get<Product[]>(`/products/public/category/${categoria}`);
    return response.data;
  },

  /**
   * Busca productos por nombre
   */
  async searchProducts(query: string): Promise<Product[]> {
    const response = await api.get<Product[]>(`/products/public/search?q=${query}`);
    return response.data;
  },
};

File: backend/src/main/java/com/pasteleria/backend/services/ProductService.java
package com.pasteleria.backend.services;

import com.pasteleria.backend.models.Product;
import com.pasteleria.backend.repositories.ProductRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

/**
 * Servicio de productos
 * Maneja la lógica de negocio relacionada con el catálogo de productos
 */
@Service
public class ProductService {

    @Autowired
    private ProductRepository productRepository;

    /**
     * Obtiene todos los productos
     */
    public List<Product> getAllProducts() {
        return productRepository.findAll();
    }

    /**
     * Obtiene solo productos disponibles
     */
    public List<Product> getAvailableProducts() {
        return productRepository.findByDisponibleTrue();
    }

    /**
     * Obtiene un producto por ID
     */
    public Product getProductById(Long id) {
        return productRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Producto no encontrado con id: " + id));
    }

    /**
     * Crea un nuevo producto (Solo ADMIN)
     */
    @Transactional
    public Product createProduct(Product product) {
        return productRepository.save(product);
    }

    /**
     * Actualiza un producto existente (Solo ADMIN)
     */
    @Transactional
    public Product updateProduct(Long id, Product productDetails) {
        Product product = getProductById(id);

        product.setNombre(productDetails.getNombre());
        product.setDescripcion(productDetails.getDescripcion());
        product.setPrecio(productDetails.getPrecio());
        product.setImagenUrl(productDetails.getImagenUrl());
        product.setCategoria(productDetails.getCategoria());
        product.setDisponible(productDetails.getDisponible());

        return productRepository.save(product);
    }

    /**
     * Elimina un producto (Solo ADMIN)
     */
    @Transactional
    public void deleteProduct(Long id) {
        Product product = getProductById(id);
        productRepository.delete(product);
    }

    /**
     * Busca productos por categoría
     */
    public List<Product> getProductsByCategory(String categoria) {
        return productRepository.findByCategoriaAndDisponibleTrue(categoria);
    }

    /**
     * Busca productos por nombre
     */
    public List<Product> searchProductsByName(String nombre) {
        return productRepository.findByNombreContainingIgnoreCase(nombre);
    }

    /**
     * Cambia la disponibilidad de un producto
     */
    @Transactional
    public Product toggleAvailability(Long id) {
        Product product = getProductById(id);
        product.setDisponible(!product.getDisponible());
        return productRepository.save(product);
    }
}